File: rust-tools/core/src/protocol.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::{Value};
use std::path::{PathBuf, Path};
use home::home_dir;

#[derive(Serialize, Deserialize, Debug)]
pub struct Command {
    pub name: String,
    pub cid: String,
    pub args: Value,
    pub shape: Value,
    pub info: bool,
    pub errors: bool,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Response {
    pub cid: String,
    pub data: Value,
}

pub type CommandHandler = fn(command: Command) -> Value;
pub type Commands = HashMap<String, CommandHandler>;

pub fn get_socket_path(name: &String) -> PathBuf {
    return home_dir().unwrap().join(
        Path::new(format!(".syml/sockets/{}.sock", name).as_str())
    );
}

File: go-tools/core/protocol.go
package core

type Data = map[string]interface{}

type Action struct {
	Cid   string `json:"cid"`
	Name  string `json:"name"`
	Args  Data   `json:"args"`
	Shape Data   `json:"shape"`

	Info   bool `json:"info"`
	Errors bool `json:"errors"`
}

type Response struct {
	Cid  string `json:"cid"`
	Data Data   `json:"data"`
}

type Callback = func(action Action) (Data, error)

var Commands = make(map[string]Callback)
